EIPS/eip-3772.md

```diff
@@ -0,0 +1,287 @@
+
+```solidity
+// uses 3 slots
+struct UserData1 {
+    uint64 amountCompressed;
+    bytes32 hash;
+    address beneficiary;
+}
+
+// uses 2 slots
+struct UserData2 {
+    uint64 amountCompressed;
+    address beneficiary;
+    bytes32 hash;
+}
+```
+
+## Backwards Compatibility
+
+There are no known backward-incompatible issues.
+
+## Reference Implementation
+
+Using structs for `cintx` as of Solidity 0.8.6 does not yet serve the purpose of saving gas, (see [solidity#11691](https://github.com/ethereum/solidity/issues/11691), requires some backward-incompatible changes in storage layout in Solidity). Hence on the implementation level `uint64` can be used directly.
+
+```soldity
+function compress(uint256 full) public pure returns (uint64 cint) {
+    uint8 bits = mostSignificantBitPosition(full);
+    if (bits <= 55) {
+        cint = uint64(full) << 8;
+    } else {
+        bits -= 55;
+        cint = (uint64(full >> bits) << 8) + bits;
+    }
+}
+
+function decompress(uint64 cint) public pure returns (uint256 full) {
+    uint8 bits = uint8(cint % (1 << 9));
+    full = uint256(cint >> 8) << bits;
+}
+
+function decompressRoundingUp(uint64 cint) public pure returns (uint256 full) {
+    uint8 bits = uint8(cint % (1 << 9));
+    full = uint256(cint >> 8) << bits + ((1 << bits) - 1);
+}
+```
+
+The above gist has `library CInt64` that contains demonstrative logic for compression, decompression, and arithmetic for `cint64`. The gist also has an example contract that uses the library for demonstration purposes.
+
+The CInt64 format is intended only for storage, while dev should convert it to uint256 form using suitable logic (decompress or decompressRoundingUp) to perform any arithmetic on it.
+
+### Some related work
+
+Some engineers have realized that using `uint256` entirely for storing value/money is not optimal in view of storage costs, and some of the bits in the slot could be better utilized with other data stored in its place.
+
+- Sushiswap's MasterChefV2: ([code ref](https://github.com/sushiswap/sushiswap/blob/master/contracts/MasterChefV2.sol#L198), they are casting `uint256` values to `uint128`, to make some room for other data. No lossy compression).
+- Dydx's Balance ([code ref](https://github.com/dydxprotocol/perpetual/blob/13eea6201cd5a89814d1f90aae2382456289c2f3/contracts/protocol/v1/lib/P1Types.sol#L45-L50), uses `uint120`, no lossy compression).
+- Compound's GovernorAlpha ([code ref](https://github.com/compound-finance/compound-protocol/blob/master/contracts/Governance/GovernorAlpha.sol#L88), they use `uint96`)
+- Rollups ([Using about 3 bytes or 24 bits for representing ETH value](https://vitalik.ca/general/2021/01/05/rollup.html#how-does-compression-work))
+
```

Comment:
EIPs are not the appropriate place for this kind of history lesson, and should not contain external links.